src/util/misc.rs

```diff
@@ -144,8 +150,9 @@ mod message_signing {
             &self,
             secp_ctx: &secp256k1::Secp256k1<C>,
             msg_hash: sha256d::Hash
-        ) -> Result<PublicKey, secp256k1::Error> {
-            let msg = secp256k1::Message::from_slice(&msg_hash[..])?;
+        ) -> Result<PublicKey, MessageSignatureError> {
+            let msg = secp256k1::Message::from_slice(&msg_hash[..])
+                .map_err(MessageSignatureError::UnableToRecoverPubKey)?;
```

Comment:
No need for the explicit `map_err` here, the `?` does the conversion for us because there is only one variant of `MessageSignatureError` with an inner secp error.

Reply:
Wouldn't an explicit map be better here? It helps differentiate between InvalidEncoding(secp256k1::Error) and UnableToRecoverPubKey(secp256k1::Error)

Reply:
There are two actually.

Reply:
`Message::from_slice` cannot fail. This should actually be an `.expect` but `?` was shorter to type and it doesn't matter anyway.

Reply:
I can add the `expect` and then remove `UnableToRecoverPublicKey` is so desired.

Reply:
What the hell, it only takes a second. Changed in 1be1bb72d1a

Reply:
Excluding the fact that I was too blind to see the second variant I cannot work out _why_ this works. If I purposefully break the call to `from_slice` the compiler returns `InvalidEncoding` even though the error conversion is ambiguous. I did not think this sort of ambiguity would get past rustc?

Reply:
@tcharding There exists a conversion from `secp256k1::Error` for `MessageSignatureError::InvalidEncoding` so this can happen implicitly since there is only one path as there was no conversion of the same error into `MessageSignatureError::UnableToRecoverPubKey`

```rust
    impl From<secp256k1::Error> for MessageSignatureError {
        fn from(e: secp256k1::Error) -> MessageSignatureError {
            MessageSignatureError::InvalidEncoding(e)
        }
    }
```

Reply:
Thanks for pointing this out and not telling me to read the code more carefully, twice in one review thread :)